src/lib/core/ReferenceCounted.h

```diff
@@ -80,4 +81,46 @@ class ReferenceCounted
     count_type mRefCount = kInitRefCount;
 };
 
+/// An variant of ReferenceCounted, with atomic counter.
+template <class Subclass, class Deletor = DeleteDeletor<Subclass>, int kInitRefCount = 1>
+class AtomicReferenceCounted
+{
+public:
+    using count_type = uint32_t;
+
+    AtomicReferenceCounted() : mRefCount(kInitRefCount) {}
+
+    /** Adds one to the usage count of this class */
+    Subclass * Retain()
+    {
+        if (mRefCount == std::numeric_limits<count_type>::max())
+        {
+            abort();
+        }
+        ++mRefCount;
+
+        return static_cast<Subclass *>(this);
+    }
+
+    /** Release usage of this class */
+    void Release()
+    {
+        if (mRefCount == 0)
+        {
+            abort();
+        }
+
+        if (--mRefCount == 0)
+        {
+            Deletor::Release(static_cast<Subclass *>(this));
+        }
```

Comment:
Similar problem here.  Two threads can see mRefCount == 1, and then two threads can decrement and underflow.

```suggestion
    switch (mRefCount--)
    {
        case 0:
            abort();
            break;
        case 1:
            Deletor::Release(static_cast<Subclass *>(this));
            break;
        default:
            break;
    }
```

Reply:
SDK is explicitly not thread-safe externally. In the ancestral Weave `Inet`, it was required internally for `AsyncDNSResolver`, which is no longer in our tree.

Reply:
Then we should get rid of the atomics.

They should either work, or not be there.  Right now they're broken.

Reply:
Now removed.  Thanks.